## House Price Prediction - Advanced regression Assignment 
### ------------------------------------------------------------------------------------------------------ 

The solution is divided into the following sections: 
- Data understanding and exploration
- Data cleaning
- Data preparation
- Model building and evaluation
- Observation and inference


In [1]:
# importing all the important
import numpy as np
import pandas as pd
pd.options.display.max_rows = 100
pd.options.display.max_columns = 100
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import linear_model
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.pipeline import Pipeline
from sklearn import metrics
import os

# hide warnings
import warnings
warnings.filterwarnings('ignore')


In [2]:
# reading the dataset
data = pd.read_csv("train.csv", encoding = 'utf-8')
data.head()

FileNotFoundError: [Errno 2] No such file or directory: 'train.csv'

In [ ]:
# Check the dimensions
data.shape

In [ ]:
# Check for column details
data.info()

In [ ]:
# To get the description of the dataset
data.describe()

In [ ]:
#checking duplicates
sum(data.duplicated(subset = 'Id')) == 0

In [ ]:
# Checking for percentage nulls
round(100*(data.isnull().sum()/len(data.index)), 2)

### Outlier Check

In [ ]:
#Checking for outlier in the numerical columns
data.describe(percentiles=[.25,.5,.75,.90,.95,.99])

### Method to remove outliers.

In [ ]:
def remove_outliers(x,y):
    q1 = x[y].quantile(0.25)
    q3 = x[y].quantile(0.75)
    value = q3-q1
    lower_value  = q1-1.5*value
    higer_value = q3+1.5*value
    out= x[(x[y]<higer_value) & (x[y]>lower_value)]
    return out

In [ ]:
#Checking the shape of the dataframe
data.shape

In [ ]:
# since, it is clear that there are multiple columns with high nulls, lets group them together
data.columns[data.isnull().any()] 

null = data.isnull().sum()/len(data)*100
null = null[null>0]
null.sort_values(inplace=True, ascending=False)
null

### According to the data dictionary provided, the nulls in these columns indicates the absence of facility which may affect the price
### Hence, we will first impute the categorical variables with 'None'


In [ ]:
# According to the data dictionary provided, the nulls in these columns indicates the absence of facility which may affect the price
# Hence, we will first impute the categorical variables with 'None'
null_with_meaning = ["Alley", "MasVnrType", "BsmtQual", "BsmtCond", "BsmtExposure", "BsmtFinType1", "BsmtFinType2", "FireplaceQu", "GarageType", "GarageFinish", "GarageQual", "GarageCond", "PoolQC", "Fence", "MiscFeature"]
for i in null_with_meaning:
    data[i].fillna("none", inplace=True)

In [ ]:
# Check nulls once again

data.columns[data.isnull().any()] 

null_2 = data.isnull().sum()/len(data)*100
null_2 = null_2[null_2>0]
null_2.sort_values(inplace=True, ascending=False)
null_2

In [ ]:
# Will check these columns one by one
data['LotFrontage'].describe()

In [ ]:
data['GarageYrBlt'].describe()

In [ ]:
data['MasVnrArea'].describe()

In [ ]:
data['Electrical'].describe()

In [ ]:
# As per the data dictionary "LotFrontage" is Linear feet of street connected to property.  
# Since it is a numeric with a fair distribution, it can be imputed with similar 'Neighborhood' values

data['LotFrontage'] = data.groupby("Neighborhood")["LotFrontage"].transform(lambda x: x.fillna(x.median()))
data["GarageYrBlt"].fillna(data["GarageYrBlt"].median(), inplace=True)
data["MasVnrArea"].fillna(data["MasVnrArea"].median(), inplace=True)
data["Electrical"].dropna(inplace=True)

In [ ]:
# Crosscheck the updated 'LotFrontage' column
data['LotFrontage'].describe()

In [ ]:
data['GarageYrBlt'].describe()

In [ ]:
data['MasVnrArea'].describe()

In [ ]:
data['Electrical'].describe()

In [ ]:
# Check the no. of rows retained
len(data.index)
len(data.index)/1460

### Some EDA on the cleaned data

#### All numeric (float and int) variables in the dataset

In [ ]:
data_numeric = data.select_dtypes(include=['float64', 'int64'])
data_numeric.head()

In [ ]:
# Dropping ID Column
data_numeric = data_numeric.drop(['Id'], axis=1)
data_numeric.head()

#### Target variable 'sale Price' vs a few select columns

In [ ]:
# plot 'Sale Price' with respect to 'Neighborhood'

plt.figure(figsize=(20, 8))
sns.barplot(x="Neighborhood", y="SalePrice", data= data)
plt.title("Sales Price with respect to Neighbourhood")
plt.xticks(rotation=90)

#### Properties in some of the Neighborhoods are high priced.

In [ ]:
# plot 'overall condition' with respect to 'Saleprice'

plt.figure(figsize=(20, 8))
sns.barplot(x="OverallCond", y="SalePrice", data= data)
plt.title("Sales Price with respect to Overall Condition")
plt.xticks(rotation=90)

In [ ]:
# plot 'overall quality' with respect to 'Saleprice'

plt.figure(figsize=(20, 8))
sns.barplot(x="OverallQual", y="SalePrice", data= data)
plt.title("Sales Price with respect to Overall Quality")
plt.xticks(rotation=90)

#### Increase in the overall quality has a direct positive effect on the sale price

In [ ]:
sns.distplot(data['SalePrice'])

In [ ]:
data_raw = data.copy

### Since the Saleprice figures are skewed towards left, we will apply the log transformation to obtain a centralized data

In [ ]:
#Log Transformation
data['SalePrice']=np.log1p(data['SalePrice'])

In [ ]:
sns.distplot(data['SalePrice'])

In [ ]:
# correlation matrix
cor = data_numeric.corr()
cor

In [ ]:
# plotting correlations on a heatmap

# figure size
plt.figure(figsize=(30,20))

# heatmap
sns.heatmap(cor, cmap="YlGnBu", annot=True)
plt.show()

#### We can see that some of the variables are correlated

#### Before dropping these columns, we will first check their predictive power

In [ ]:
# Checking the same with a pairplot 
sns.set()
cols = ['SalePrice', 'GrLivArea', 'GarageCars', 'BsmtUnfSF', 'BsmtFinSF1', 'GarageArea', 'TotalBsmtSF', 'YearBuilt', 'TotRmsAbvGrd', 'GarageYrBlt']
sns.pairplot(data[cols], size = 2.5)
plt.show()

### Drop columns that are correlated and not contributing to 'SalePrice'

In [ ]:
data = data.drop(['GarageCars'], axis = 1)
data = data.drop(['BsmtUnfSF'], axis = 1)
data = data.drop(['TotRmsAbvGrd'], axis = 1)
data = data.drop(['GarageYrBlt'], axis = 1)    

data.head()

In [ ]:
#Numeric columns
data.select_dtypes(exclude=['object'])

In [ ]:
# Analyse some important numeric columns
sns.jointplot(x='GrLivArea', y='SalePrice', data=data)
plt.show()

In [ ]:
# Removing some outliers on lower right side of 'GrLivArea'
data = remove_outliers(data,'GrLivArea')

### Since the dataset is small it isn't advisable to do remove outliers.

In [ ]:
data.shape

In [ ]:
# Again plotting GeLivArea vs SalePrice
sns.jointplot(x = data['GrLivArea'], y = data['SalePrice'])
plt.show()

In [ ]:
# Lot frontage vs SalePrice 
sns.jointplot(x = data['LotFrontage'], y = data['SalePrice'])
plt.show()

In [ ]:
# LotArea vs SalePrice
sns.jointplot(x = data['LotArea'], y = data['SalePrice'])
plt.show()

In [ ]:
# 1stFlrSF vs SalePrice
sns.jointplot(x = data['1stFlrSF'], y = data['SalePrice'])
plt.show()

In [ ]:
# 2ndFlrSF vs SalePrice
sns.jointplot(x = data['2ndFlrSF'], y = data['SalePrice'])
plt.show()

In [ ]:
# OverallQual vs SalePrice
sns.jointplot(x = data['OverallQual'], y = data['SalePrice'])
plt.show()

In [ ]:
# OverallCond vs SalePrice
sns.jointplot(x=data['OverallCond'], y = data['SalePrice'])
plt.show()

#### Ground or First level houses i.e. '0' second floor Sq.Ft has also a steady increase

### We can derive a column for 'Age of the property' when it was sold: Name it as 'PropAge'

In [ ]:
# PropAge -  Property Age from yearsold - yearbuilt
data['PropAge'] = (data['YrSold'] - data['YearBuilt'])
data.head()

In [ ]:
# PropAge vs SalePrice
sns.jointplot(x = data['PropAge'], y = data['SalePrice'])
plt.show()

#### Increase in Property Age shows a decreasing saleprice trend i.e newer the property, high is the value

###  Now we can drop the column Month sold and Year Sold, Year built and Year remodelled since it will not be required further 

In [ ]:
data = data.drop(['MoSold'], axis = 1)
data = data.drop(['YrSold'], axis = 1)
data = data.drop(['YearBuilt'], axis = 1)
data = data.drop(['YearRemodAdd'], axis = 1)
data.head()

In [ ]:
data.Street.value_counts()

In [ ]:
data.Utilities.value_counts()

In [ ]:
# We can also drop columns that show very low variance and thus not required for predictions
data = data.drop(['Street'], axis = 1)
data = data.drop(['Utilities'], axis = 1)

### Just to check the variance of these columns 

In [ ]:

# l1 = ['Condition2', 'Heating', 'PoolQC', 'RoofMatl', 'BsmtCond', 'GarageQual', 'GarageCond', 'MiscVal', '3SsnPorch', 'FireplaceQu', 'BsmtHalfBath', 'BsmtFinSF2', 'Alley', 'MiscFeature', 'Fence', 'Functional']
l2= data.select_dtypes(include=['float64', 'int64'])
l2



In [ ]:
for i in l2:
    print(data[i].value_counts())

In [ ]:
 data = data.drop(['PoolQC','MiscVal', 'Alley', 'RoofMatl', 'Condition2', 'Heating', 'GarageCond', 'Fence', 'Functional' ], axis = 1)


### These Columns were having high null values, some of which were imputed. After imputing, it was found that there was very little variance in the data. So we have decided to drop these columns.

In [ ]:
data.shape

## 3. Data Preparation 


#### Data Preparation

Let's now prepare the data and build the model.

In [ ]:
# Drop 'Id' from Dataframe

data = data.drop(['Id'], axis=1)
data.head()

In [ ]:
#type of each feature in data: int, float, object
types = data.dtypes
#numerical values are either type int or float
numeric_type = types[(types == 'int64') | (types == float)] 
#categorical values are type object
categorical_type = types[types == object]

In [ ]:
pd.DataFrame(types).reset_index().set_index(0).reset_index()[0].value_counts()

In [ ]:
#we should convert numeric_type to a list to make it easier to work with
numerical_columns = list(numeric_type.index)
print(numerical_columns)

In [ ]:
#Categorical columns
categorical_columns = list(categorical_type.index)
print(categorical_columns)

### Creating Dummy columns to convert categorical into numerical

In [ ]:
data = pd.get_dummies(data, drop_first=True )
data.head()

In [ ]:
X = data.drop(['SalePrice'], axis=1)

X.head()

In [3]:
# Putting response variable to y
y = data['SalePrice']

y.head()

NameError: name 'data' is not defined

In [4]:
# Splitting the data into train and test
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, test_size=0.3, random_state=50)

NameError: name 'X' is not defined

In [5]:
from sklearn.preprocessing import StandardScaler

In [6]:
scaler = StandardScaler()

X_train[['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'Fireplaces', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'PropAge']] = scaler.fit_transform(X_train[['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'Fireplaces', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'PropAge']])

X_test[['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'Fireplaces', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'PropAge']] = scaler.fit_transform(X_test[['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'Fireplaces', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'PropAge']])



NameError: name 'X_train' is not defined

In [7]:
X_train.head()

NameError: name 'X_train' is not defined

In [8]:
X_test.head()

NameError: name 'X_test' is not defined

## 3. Model Building and Evaluation

## Lets first check the model using Linear Regression and RFE (OPTIONAL)

In [9]:
# Importing RFE and LinearRegression
# Since there are more variables to be analysed, we will used the automated feature elimination process (RFE)

from sklearn.feature_selection import RFE
from sklearn.linear_model import LinearRegression
import statsmodels.api as sm

In [10]:
# Running RFE 
# Since there are more than 250 variables for analysis, we will run RFE to select some that have high predictive power
lm = LinearRegression()
lm.fit(X_train, y_train)
 # running RFE for top 100 variables
rfe = RFE(lm, 100)            
rfe = rfe.fit(X_train, y_train)

NameError: name 'X_train' is not defined

In [ ]:
# Check the ranks
list(zip(X_train.columns,rfe.support_,rfe.ranking_))

In [ ]:
# Select the top 100 variables

col = X_train.columns[rfe.support_]
col

In [ ]:
X_train.columns[~rfe.support_]

In [ ]:
# Creating X_test dataframe with RFE selected variables
X_train_rfe = X_train[col]

In [ ]:
X_train_rfe = pd.DataFrame(X_train[col])

In [ ]:
X_train_rfe.head()

In [11]:
X_train_rfe.shape

NameError: name 'X_train_rfe' is not defined

In [12]:
# predict
y_train_pred = lm.predict(X_train)
metrics.r2_score(y_true=y_train, y_pred=y_train_pred)

NameError: name 'X_train' is not defined

In [13]:
y_test_pred = lm.predict(X_test)
metrics.r2_score(y_true=y_test, y_pred=y_test_pred)

NameError: name 'X_test' is not defined

## Since the Test R2 is too low, we will check for some alternate methods of Regression

In [14]:
# Check the ranks
list(zip(X_test.columns,rfe.support_,rfe.ranking_))

NameError: name 'X_test' is not defined

In [15]:
# Select the top 100 variables

col1 = X_test.columns[rfe.support_]
col1

NameError: name 'X_test' is not defined

In [16]:
X_test_rfe = X_test[col1]

NameError: name 'X_test' is not defined

In [17]:
X_test_rfe.head()

NameError: name 'X_test_rfe' is not defined

##  Lasso and Ridge Regression

Let's now try predicting house prices and perform lasso and ridge regression.

### Lasso Regression

In [18]:
# Checking the dimension of X_train & y_train
print("X_train", X_train.shape)
print("y_train", y_train.shape)

NameError: name 'X_train' is not defined

In [19]:
# Applying Lasso

# list of alphas to tune
params = {'alpha': [0.0001, 0.001, 0.01, 0.05, 0.1, 
 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 2.0, 3.0, 
 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 20, 50, 100, 500, 1000 ]}
lasso = Lasso()

# cross validation
folds = 5
model_cv = GridSearchCV(estimator = lasso, 
                        param_grid = params, 
                        scoring= 'neg_mean_absolute_error', 
                        cv = folds, 
                        return_train_score=True,
                        verbose = 1)            

model_cv.fit(X_train, y_train) 

NameError: name 'X_train' is not defined

In [20]:
# cv_results
cv_results = pd.DataFrame(model_cv.cv_results_)
cv_results = cv_results[cv_results['param_alpha']<=1]
cv_results.head()

AttributeError: 'GridSearchCV' object has no attribute 'cv_results_'

In [21]:
# plotting mean test and train scoes with alpha 
cv_results['param_alpha'] = cv_results['param_alpha'].astype('float32')

# plotting
plt.plot(cv_results['param_alpha'], cv_results['mean_train_score'])
plt.plot(cv_results['param_alpha'], cv_results['mean_test_score'])
plt.xlabel('alpha')
plt.ylabel('Negative Mean Absolute Error')
plt.title("Negative Mean Absolute Error and alpha")
plt.legend(['train score', 'test score'], loc='upper left')
plt.show()

NameError: name 'cv_results' is not defined

#### From the above graph we can see that the Negative Mean Absolute Error is quite low at alpha = 0.4 and stabilises thereafter, 
#### but we will choose a low value of alpha to balance the trade-off between Bias-Variance 
#### and to get the coefficients of smallest of features.

In [22]:
# At alpha = 0.01, even the smallest of negative coefficients that have some predictive power towards 'SalePrice' have been generated

alpha = 0.01
lasso = Lasso(alpha=alpha)

lasso.fit(X_train, y_train)
lasso.coef_

NameError: name 'X_train' is not defined

#### The advantage of this technique is clearly visible here as Lasso brings the coefficients of insignificant features to zero

In [23]:
# lasso model parameters
model_parameters = list(lasso.coef_ )
model_parameters.insert(0, lasso.intercept_)
model_parameters = [round(x, 3) for x in model_parameters]
cols = X.columns
cols = cols.insert(0, "constant")
list(zip(cols, model_parameters))

AttributeError: 'Lasso' object has no attribute 'coef_'

In [24]:
# lasso regression
lm = Lasso(alpha=0.01)
lm.fit(X_train, y_train)

# prediction on the test set(Using R2)
y_train_pred = lm.predict(X_train)
print(metrics.r2_score(y_true=y_train, y_pred=y_train_pred))
y_test_pred = lm.predict(X_test)
print(metrics.r2_score(y_true=y_test, y_pred=y_test_pred))

NameError: name 'X_train' is not defined

In [25]:
print('RMSE :', np.sqrt(metrics.mean_squared_error(y_test, y_test_pred)))

NameError: name 'y_test' is not defined

In [26]:
#### The R2 values for Train and Test matches well, indicating an optimum model

In [27]:
# Creating a dataframe for the coefficients obtained from Lasso
mod = list(zip(cols, model_parameters))

NameError: name 'cols' is not defined

In [28]:
para = pd.DataFrame(mod)
para.columns = ['Variable', 'Coeff']
para.head()

NameError: name 'mod' is not defined

In [29]:
# sort the coefficients in ascending order
para = para.sort_values((['Coeff']), axis = 0, ascending = False)
para

NameError: name 'para' is not defined

In [30]:
# Chose variables whose coefficients are non-zero
pred = pd.DataFrame(para[(para['Coeff'] != 0)])
pred

NameError: name 'para' is not defined

In [31]:
# These 16 variables obtained from Lasso Regression can be concluded to have the strong effect on the SalePrice
pred.shape

NameError: name 'pred' is not defined

In [32]:
Lassso_var = list(pred['Variable'])
print(Lassso_var)

NameError: name 'pred' is not defined

In [33]:
X_train_lasso = X_train[['GrLivArea', 'OverallQual', 'OverallCond', 'TotalBsmtSF', 'GarageArea', 'BsmtFinSF1', 'Fireplaces', 'LotArea', 'LotFrontage', 'BsmtFullBath', 'Foundation_PConc', 'OpenPorchSF', 'FullBath', 'ScreenPorch', 'WoodDeckSF']]
                        
X_train_lasso.head()

NameError: name 'X_train' is not defined

In [34]:
X_train_lasso.shape

NameError: name 'X_train_lasso' is not defined

In [35]:
X_test_lasso = X_test[['GrLivArea', 'OverallQual', 'OverallCond', 'TotalBsmtSF', 'GarageArea', 'BsmtFinSF1', 'Fireplaces', 'LotArea', 'LotFrontage', 'BsmtFullBath', 'Foundation_PConc', 'OpenPorchSF', 'FullBath', 'ScreenPorch', 'WoodDeckSF']]
                        
X_test_lasso.head()

NameError: name 'X_test' is not defined

## Ridge Regression

In [36]:
# list of alphas to tune
params = {'alpha': [0.0001, 0.001, 0.01, 0.05, 0.1, 
 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 2.0, 3.0, 
 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 20, 50, 100, 500, 1000 ]}


ridge = Ridge()

# cross validation
folds = 5
model_cv = GridSearchCV(estimator = ridge, 
                        param_grid = params, 
                        scoring= 'neg_mean_absolute_error', 
                        cv = folds, 
                        return_train_score=True,
                        verbose = 1)            
model_cv.fit(X_train, y_train) 

NameError: name 'X_train' is not defined

In [37]:
cv_results = pd.DataFrame(model_cv.cv_results_)
cv_results = cv_results[cv_results['param_alpha']<=5]
cv_results.head()

AttributeError: 'GridSearchCV' object has no attribute 'cv_results_'

In [38]:
# plotting mean test and train scoes with alpha 
cv_results['param_alpha'] = cv_results['param_alpha'].astype('int32')

# plotting
plt.plot(cv_results['param_alpha'], cv_results['mean_train_score'])
plt.plot(cv_results['param_alpha'], cv_results['mean_test_score'])
plt.xlabel('alpha')
plt.ylabel('Negative Mean Absolute Error')
plt.title("Negative Mean Absolute Error and alpha")
plt.legend(['train score', 'test score'], loc='upper left')
plt.show()

NameError: name 'cv_results' is not defined

### since the Negative Mean Absolute Error stabilises at alpha = 2, we will choose this for further analysis

In [107]:
alpha = 2
ridge = Ridge(alpha=alpha)

ridge.fit(X_train, y_train)
ridge.coef_

array([-0.01661814,  0.00879344,  0.02534461,  0.06457657,  0.05124942,
       -0.00169877,  0.0304906 ,  0.00747482,  0.03650081,  0.03991915,
        0.04767425,  0.00523671,  0.07595503,  0.01170416, -0.00092158,
        0.00892371,  0.01249545,  0.0003776 , -0.00534099,  0.00350875,
        0.02078896,  0.01348507, -0.0010637 ,  0.01230575,  0.00491464,
        0.0141215 ,  0.00971184, -0.0702245 ,  0.14857807,  0.10523788,
        0.12502615,  0.09704995,  0.00219378, -0.0208621 ,  0.00027408,
        0.01235864, -0.02072982,  0.00302972,  0.01744635, -0.02148039,
       -0.05117096, -0.02397809,  0.01020923, -0.03899013, -0.0150698 ,
       -0.02340803,  0.00931668,  0.03728476, -0.00759928,  0.11394039,
       -0.05593223, -0.00971649, -0.08597375, -0.11266515, -0.04195278,
       -0.01867322,  0.04765164, -0.02093394,  0.03560318,  0.04414944,
       -0.0527455 , -0.04663915, -0.00821161,  0.00358752,  0.02349391,
        0.09259274, -0.00371216,  0.03744706,  0.01443113,  0.06

In [108]:
# ridge model parameters
model_parameters = list(ridge.coef_)
model_parameters.insert(0, ridge.intercept_)
model_parameters = [round(x, 3) for x in model_parameters]
cols = X.columns
cols = cols.insert(0, "constant")
list(zip(cols, model_parameters))

[('constant', 11.739),
 ('MSSubClass', -0.017),
 ('LotFrontage', 0.009),
 ('LotArea', 0.025),
 ('OverallQual', 0.065),
 ('OverallCond', 0.051),
 ('MasVnrArea', -0.002),
 ('BsmtFinSF1', 0.03),
 ('BsmtFinSF2', 0.007),
 ('TotalBsmtSF', 0.037),
 ('1stFlrSF', 0.04),
 ('2ndFlrSF', 0.048),
 ('LowQualFinSF', 0.005),
 ('GrLivArea', 0.076),
 ('BsmtFullBath', 0.012),
 ('BsmtHalfBath', -0.001),
 ('FullBath', 0.009),
 ('HalfBath', 0.012),
 ('BedroomAbvGr', 0.0),
 ('KitchenAbvGr', -0.005),
 ('Fireplaces', 0.004),
 ('GarageArea', 0.021),
 ('WoodDeckSF', 0.013),
 ('OpenPorchSF', -0.001),
 ('EnclosedPorch', 0.012),
 ('3SsnPorch', 0.005),
 ('ScreenPorch', 0.014),
 ('PoolArea', 0.01),
 ('PropAge', -0.07),
 ('MSZoning_FV', 0.149),
 ('MSZoning_RH', 0.105),
 ('MSZoning_RL', 0.125),
 ('MSZoning_RM', 0.097),
 ('LotShape_IR2', 0.002),
 ('LotShape_IR3', -0.021),
 ('LotShape_Reg', 0.0),
 ('LandContour_HLS', 0.012),
 ('LandContour_Low', -0.021),
 ('LandContour_Lvl', 0.003),
 ('LotConfig_CulDSac', 0.017),
 ('LotCo

In [109]:
# ridge regression
lm = Ridge(alpha=2)
lm.fit(X_train, y_train)

# predict
y_train_pred = lm.predict(X_train)
print(metrics.r2_score(y_true=y_train, y_pred=y_train_pred))
y_test_pred = lm.predict(X_test)
print(metrics.r2_score(y_true=y_test, y_pred=y_test_pred))

0.9364594823911133
0.9077597079466581


In [110]:
print('RMSE :', np.sqrt(metrics.mean_squared_error(y_test, y_test_pred)))

RMSE : 0.11485785595060997


#### It is visible that the model performance is better than Lasso.  The train and the test scores are matching well

In [111]:
# Create a dataframe for Ridge Coefficients
mod_ridge = list(zip(cols, model_parameters))

In [112]:
paraRFE = pd.DataFrame(mod_ridge)
paraRFE.columns = ['Variable', 'Coeff']
res=paraRFE.sort_values(by=['Coeff'], ascending = False)
res.head(20)

Variable   Coeff
0                 constant  11.739
29             MSZoning_FV   0.149
31             MSZoning_RL   0.125
50    Neighborhood_Crawfor   0.114
30             MSZoning_RH   0.105
32             MSZoning_RM   0.097
210  SaleCondition_Partial   0.097
66    Neighborhood_StoneBr   0.093
13               GrLivArea   0.076
209   SaleCondition_Normal   0.074
95     Exterior1st_BrkFace   0.069
70         Condition1_Norm   0.067
136       Foundation_Stone   0.067
4              OverallQual   0.065
206  SaleCondition_AdjLand   0.061
200         SaleType_ConLD   0.058
103     Exterior1st_Stucco   0.055
198           SaleType_CWD   0.054
124       MasVnrType_Stone   0.052
5              OverallCond   0.051

In [113]:
# Sorting the coefficients in ascending order
paraRFE = paraRFE.sort_values((['Coeff']), axis = 0, ascending = False)
paraRFE

Variable   Coeff
0                 constant  11.739
29             MSZoning_FV   0.149
31             MSZoning_RL   0.125
50    Neighborhood_Crawfor   0.114
30             MSZoning_RH   0.105
32             MSZoning_RM   0.097
210  SaleCondition_Partial   0.097
66    Neighborhood_StoneBr   0.093
13               GrLivArea   0.076
209   SaleCondition_Normal   0.074
95     Exterior1st_BrkFace   0.069
70         Condition1_Norm   0.067
136       Foundation_Stone   0.067
4              OverallQual   0.065
206  SaleCondition_AdjLand   0.061
200         SaleType_ConLD   0.058
103     Exterior1st_Stucco   0.055
198           SaleType_CWD   0.054
124       MasVnrType_Stone   0.052
5              OverallCond   0.051
134       Foundation_PConc   0.051
166           CentralAir_Y   0.051
176         FireplaceQu_Po   0.050
74         Condition1_RRAn   0.049
11                2ndFlrSF   0.048
57    Neighborhood_NPkVill   0.048
72         Condition1_PosN   0.048
146        BsmtExposure_Gd   0.048
60    Neighborhood_NridgHt   0.044
183      GarageType_Detchd   0.041
179      GarageType_Attchd   0.040
10                1stFlrSF   0.040
91       RoofStyle_Mansard   0.039
120    Exterior2nd_Wd Sdng   0.039
9              TotalBsmtSF   0.037
68    Neighborhood_Veenker   0.037
48    Neighborhood_ClearCr   0.037
59    Neighborhood_NoRidge   0.036
199           SaleType_Con   0.033
77         BldgType_2fmCon   0.033
111    Exterior2nd_CmentBd   0.033
122     MasVnrType_BrkFace   0.030
7               BsmtFinSF1   0.030
189          GarageQual_Gd   0.029
76         Condition1_RRNn   0.028
181     GarageType_BuiltIn   0.028
100    Exterior1st_MetalSd   0.027
197       MiscFeature_none   0.026
3                  LotArea   0.025
123        MasVnrType_None   0.024
..                     ...     ...
178       FireplaceQu_none  -0.019
56      Neighborhood_NAmes  -0.019
161      BsmtFinType2_none  -0.020
34            LotShape_IR3  -0.021
37         LandContour_Low  -0.021
40           LotConfig_FR2  -0.021
58     Neighborhood_NWAmes  -0.021
97     Exterior1st_CemntBd  -0.021
162           HeatingQC_Fa  -0.021
98     Exterior1st_HdBoard  -0.022
84       HouseStyle_2.5Unf  -0.022
104    Exterior1st_VinylSd  -0.023
46     Neighborhood_BrDale  -0.023
42        LotConfig_Inside  -0.024
153       BsmtFinType1_Rec  -0.024
88         RoofStyle_Gable  -0.025
174         FireplaceQu_Fa  -0.025
113    Exterior2nd_ImStucc  -0.028
152       BsmtFinType1_LwQ  -0.029
118     Exterior2nd_Stucco  -0.030
165           HeatingQC_TA  -0.033
169         Electrical_Mix  -0.034
71         Condition1_PosA  -0.034
90           RoofStyle_Hip  -0.034
109    Exterior2nd_BrkFace  -0.035
44           LandSlope_Sev  -0.039
117      Exterior2nd_Stone  -0.041
190          GarageQual_Po  -0.041
55    Neighborhood_Mitchel  -0.042
188          GarageQual_Fa  -0.043
159       BsmtFinType2_Rec  -0.044
105    Exterior1st_Wd Sdng  -0.046
62      Neighborhood_SWISU  -0.047
129           ExterCond_Fa  -0.048
89       RoofStyle_Gambrel  -0.049
79          BldgType_Twnhs  -0.049
108    Exterior2nd_Brk Cmn  -0.050
41           LotConfig_FR3  -0.051
61    Neighborhood_OldTown  -0.053
51    Neighborhood_Edwards  -0.056
73         Condition1_RRAe  -0.056
137        Foundation_Wood  -0.059
172         KitchenQual_Gd  -0.065
171         KitchenQual_Fa  -0.068
28                 PropAge  -0.070
173         KitchenQual_TA  -0.080
53     Neighborhood_IDOTRR  -0.086
94     Exterior1st_BrkComm  -0.096
83       HouseStyle_2.5Fin  -0.100
54    Neighborhood_MeadowV  -0.113

[211 rows x 2 columns]

In [114]:
## since there were few coefficients at 0, we removed them from features
predRFE = pd.DataFrame(paraRFE[(paraRFE['Coeff'] != 0)])
predRFE

Variable   Coeff
0                 constant  11.739
29             MSZoning_FV   0.149
31             MSZoning_RL   0.125
50    Neighborhood_Crawfor   0.114
30             MSZoning_RH   0.105
32             MSZoning_RM   0.097
210  SaleCondition_Partial   0.097
66    Neighborhood_StoneBr   0.093
13               GrLivArea   0.076
209   SaleCondition_Normal   0.074
95     Exterior1st_BrkFace   0.069
70         Condition1_Norm   0.067
136       Foundation_Stone   0.067
4              OverallQual   0.065
206  SaleCondition_AdjLand   0.061
200         SaleType_ConLD   0.058
103     Exterior1st_Stucco   0.055
198           SaleType_CWD   0.054
124       MasVnrType_Stone   0.052
5              OverallCond   0.051
134       Foundation_PConc   0.051
166           CentralAir_Y   0.051
176         FireplaceQu_Po   0.050
74         Condition1_RRAn   0.049
11                2ndFlrSF   0.048
57    Neighborhood_NPkVill   0.048
72         Condition1_PosN   0.048
146        BsmtExposure_Gd   0.048
60    Neighborhood_NridgHt   0.044
183      GarageType_Detchd   0.041
179      GarageType_Attchd   0.040
10                1stFlrSF   0.040
91       RoofStyle_Mansard   0.039
120    Exterior2nd_Wd Sdng   0.039
9              TotalBsmtSF   0.037
68    Neighborhood_Veenker   0.037
48    Neighborhood_ClearCr   0.037
59    Neighborhood_NoRidge   0.036
199           SaleType_Con   0.033
77         BldgType_2fmCon   0.033
111    Exterior2nd_CmentBd   0.033
122     MasVnrType_BrkFace   0.030
7               BsmtFinSF1   0.030
189          GarageQual_Gd   0.029
76         Condition1_RRNn   0.028
181     GarageType_BuiltIn   0.028
100    Exterior1st_MetalSd   0.027
197       MiscFeature_none   0.026
3                  LotArea   0.025
123        MasVnrType_None   0.024
..                     ...     ...
178       FireplaceQu_none  -0.019
56      Neighborhood_NAmes  -0.019
161      BsmtFinType2_none  -0.020
34            LotShape_IR3  -0.021
37         LandContour_Low  -0.021
40           LotConfig_FR2  -0.021
58     Neighborhood_NWAmes  -0.021
97     Exterior1st_CemntBd  -0.021
162           HeatingQC_Fa  -0.021
98     Exterior1st_HdBoard  -0.022
84       HouseStyle_2.5Unf  -0.022
104    Exterior1st_VinylSd  -0.023
46     Neighborhood_BrDale  -0.023
42        LotConfig_Inside  -0.024
153       BsmtFinType1_Rec  -0.024
88         RoofStyle_Gable  -0.025
174         FireplaceQu_Fa  -0.025
113    Exterior2nd_ImStucc  -0.028
152       BsmtFinType1_LwQ  -0.029
118     Exterior2nd_Stucco  -0.030
165           HeatingQC_TA  -0.033
169         Electrical_Mix  -0.034
71         Condition1_PosA  -0.034
90           RoofStyle_Hip  -0.034
109    Exterior2nd_BrkFace  -0.035
44           LandSlope_Sev  -0.039
117      Exterior2nd_Stone  -0.041
190          GarageQual_Po  -0.041
55    Neighborhood_Mitchel  -0.042
188          GarageQual_Fa  -0.043
159       BsmtFinType2_Rec  -0.044
105    Exterior1st_Wd Sdng  -0.046
62      Neighborhood_SWISU  -0.047
129           ExterCond_Fa  -0.048
89       RoofStyle_Gambrel  -0.049
79          BldgType_Twnhs  -0.049
108    Exterior2nd_Brk Cmn  -0.050
41           LotConfig_FR3  -0.051
61    Neighborhood_OldTown  -0.053
51    Neighborhood_Edwards  -0.056
73         Condition1_RRAe  -0.056
137        Foundation_Wood  -0.059
172         KitchenQual_Gd  -0.065
171         KitchenQual_Fa  -0.068
28                 PropAge  -0.070
173         KitchenQual_TA  -0.080
53     Neighborhood_IDOTRR  -0.086
94     Exterior1st_BrkComm  -0.096
83       HouseStyle_2.5Fin  -0.100
54    Neighborhood_MeadowV  -0.113

[204 rows x 2 columns]

In [115]:
predRFE.shape

(204, 2)

### Observation:
#### Though the model performance by Ridge Regression was better in terms of R2 values of Train and Test, 
#### it is better to use Lasso, since it brings and assigns a zero value to insignificant features, enabling us to choose
#### the predictive variables.
#### It is always advisable to use simple yet robust model.

#### Equation can be formulated using the features and coefficients obtained by Lasso

In [116]:
### Assign the Features as x1, x2.....

pred.set_index(pd.Index(['C','x1', 'x2', 'x3', 'x4', 'x5' , 'x6', 'x7', 'x8', 'x9', 'x10', 'x11', 'x12', 'x13', 'x14', 'x15', 'x16']), inplace = True) 
pred

Variable   Coeff
C          constant  12.003
x1        GrLivArea   0.125
x2      OverallQual   0.112
x3      OverallCond   0.050
x4      TotalBsmtSF   0.042
x5       BsmtFinSF1   0.035
x6       GarageArea   0.034
x7       Fireplaces   0.024
x8          LotArea   0.015
x9      LotFrontage   0.014
x10    BsmtFullBath   0.010
x11      WoodDeckSF   0.010
x12     ScreenPorch   0.005
x13  KitchenQual_TA  -0.007
x14      MSSubClass  -0.007
x15    KitchenAbvGr  -0.008
x16         PropAge  -0.095

### These are the final features that should be selected for predicting the price of house

### Hence the equation:
    
####    Log(Y) = C + 0.125(x1) + 0.112(x2) +  0.050(x3) + 0.042(x4) + 0.035(x5) + 0.034(x6) + 0.024(x7) +  0.015(x8) + 0.014(x9) + 0.010(x10)
####                + 0.010(x11) + 0.005(x12) - 0.007(x13) - 0.007(x14) - 0.008(x15) - 0.095(x16) + Error term(RSS + alpha * (sum of absolute value of coefficients)

### INFERENCE


#### Suggestions for Surprise Housing is to keep a check on these predictors affecting the price of the house.

#### The higher values of positive coeeficients suggest a high sale value.

#### Some of those features are:-
 |  Feature  |  Description  |
 |  ---  |  ---  |
 |  GrLivArea  |  Above grade (ground) living area square feet  |
 |  OverallQual  |  Rates the overall material and finish of the house  |
 |  OverallCond  |  Rates the overall condition of the house  |
 |  TotalBsmtSF  |  Total square feet of basement area  |
 |  GarageArea   |Size of garage in square feet  |
        
#### The higher values of negative coeeficients suggest a decrease in sale value.

#### Some of those features are:-
   |  Feature  |  Description  |
   |  ---  |  ---  |
   |  PropAge  |  Age of the property at the time of seeling  |
   |  MSSubClass  |  Identifies the type of dwelling involved in the sale  |
    

#### When the market value of the property is lower than the Predicted Sale Price, its the time to buy.

